crates/circuit/src/register.rs

```diff
@@ -0,0 +1,1054 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::{
+    bit::{
+        BitExtraInfo, BitInfo, PyAncillaQubit, PyBit, PyClbit, PyQubit, ShareableClbit,
+        ShareableQubit,
+    },
+    circuit_data::CircuitError,
+    slice::{PySequenceIndex, PySequenceIndexError},
+};
+use indexmap::IndexSet;
+use pyo3::{
+    exceptions::PyValueError,
+    prelude::*,
+    types::{PyList, PyString, PyTuple, PyType},
+    IntoPyObjectExt, PyTypeInfo,
+};
+use std::{
+    cmp::Eq,
+    fmt::Display,
+    hash::Hash,
+    sync::{atomic::AtomicU32, Arc},
+};
+
+pub trait Register {
+    /// The type of bit stored by the [Register]
+    type Bit;
+
+    /// Returns the size of the [Register].
+    fn len(&self) -> usize;
+    /// Checks if the [Register] is empty.
+    fn is_empty(&self) -> bool;
+    /// Returns the name of the [Register].
+    fn name(&self) -> &str;
+    /// Checks if a bit exists within the [Register].
+    fn contains(&self, bit: &Self::Bit) -> bool;
+    /// Return an iterator over all the bits in the register
+    fn bits(&self) -> impl ExactSizeIterator<Item = Self::Bit>;
+    /// Gets a bit by index
+    fn get(&self, index: usize) -> Option<Self::Bit>;
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(crate) enum RegisterInfo {
+    Owning(Arc<OwningRegisterInfo>),
+    Alias {
+        name: String,
+        bits: Box<IndexSet<BitInfo>>,
```

Comment:
I would make this:

```suggestion
        bits: Box<IndexSet<BitInfo, ahash::RandomState>>,
```

to speedup lookups.

Reply:
Obsolete in 883d82334.